book/src/dev/rfcs/0005-treestate.md

```diff
@@ -0,0 +1,194 @@
+Sapling anchors). Sapling block validation includes comparing the specified
+FinalSaplingRoot in its block header to the root of the Sapling note commitment
+tree that we have just computed to make sure they match.
+
+As the transactions within a block are parsed, Sapling shielded transactions
+including Spend descriptions and Output descriptions describe the spending and
+creation of Zcash Sapling notes, and JoinSplit-on-Groth16 descriptions to
+transfer/spend/create Sprout notes and transparent value. JoinSplit and Spend
+descriptions specify an anchor, which references a previous note commitment tree
+root: for Spend's, this is a previous block's anchor as defined in their block
+header, for JoinSplit's, it may be a previous block's anchor or the root
+produced by a strictly previous JoinSplit description in its transaction. For
+Spend's, this is convenient because we can query our state service for
+previously finalized Sapling block anchors, and if they are found, then that
+[consensus check](https://zips.z.cash/protocol/canopy.pdf#spendsandoutputs) has
+been satisfied and the Spend description can be validated independently. For
+JoinSplit's, if it's not a previously finalized block anchor, it must be the
+treestate anchor of previous JoinSplit in this transaction, and we have to wait
+for that one to be parsed and its root computed to check that ours is
+valid. Luckily, it can only be a previous JoinSplit in this transaction, and are
+[usually the immediately previous one](zcashd), so the set of candidate anchors
+is smaller for earlier JoinSplit's in a transcation, but larger for the later
+ones. For these JoinSplit's, they can be validated independently of their
+anchor's finalization status as long as the final check of the anchor is done,
+when available, such as at the Transaction level after all the JoinSplit's have
+finished validating everything that can be validated without the context of
+their anchor's finalization state.
+
+So for each transaction, for both Spend descriptions and JoinSplit's, we can
+pre-emptively try to do our consensus check by looking up the anchors in our
+finalized that first. For Spend's, we then trigger the remaining validation and
+when that finishes we are full done with those. For JoinSplit's, the anchor
+state check may pass early if it's a previous block Sprout note commitment tree
+root, but it may fail because it's an earlier JoinSplit's root instead, so once
+the JoinSplit validates independently of the anchor, we wait for all candidate
+previous JoinSplit's in that transaction finish validating before doing the
+anchor consensus check again, but against the output treestate roots of earlier
+JoinSplit's.
+
+Both Sprout and Sapling note commitment trees must be computed for the whole
+block to validate, but Sprout, we need to compute interstitial treestates in
+between JoinSplit's in order to do the final consensus check for each/all
+JoinSplit's, not just for the whole block, like in Sapling.
```

Comment:
Consider splitting this sentence into the different sprout and sapling checks. I had to read it a few times to understand it.

Reply:
https://github.com/ZcashFoundation/zebra/pull/983/commits/880e85e23c64ec669808067a3e034837626b0c26